Make non-ff updates work for "unsafe branches"

forslund opened the pull request:
Unsafe branches are currently only "master"

The method is:
- find branch of sha
- checkout that branch
- reset to a point where the merge will work even if the head of the current branch is ahead of the sha
- perform merge as normal

forslund committed:
Make non-ff updates work for "unsafe branches"

Unsafe branches are currently only "master"

MatthewScholefield commented on msm/skill_entry.py:
```diff
@@ -208,6 +211,19 @@ def update(self):
+                git.checkout(sha_branch)
+                # Reset to common ancestor to make the merge work OK
+                common_sha = git.merge_base(sha_before, self.sha)
+                git.reset(common_sha, hard=True)
```
Any reason we need to hard reset once we've switched branches? This would delete any new commits on `18.02`. ie.:
```
git checkout 18.02
git commit -a -m "Add feature"
git checkout master
msm update
# Now the "Add feature" commit is lost
```

MatthewScholefield committed:
Handle sha only existing in remote branch
This also removes the hard reset which isn't necessary after switching branches

MatthewScholefield committed:
Move modified file checking out of branch switching code
It makes sense to have only one way to check if the skill is modified
Also, this prevents SkillModified from being thrown too frequently

MatthewScholefield committed:
Rename UNSAFE_BRANCHES to SWITCHABLE_BRANCHES
Now with the new behavior, it's not necessarily unsafe but rather prone to being switched off of

MatthewScholefield committed:
Handle an empty sha when updating skills

forslund commented:
Looks good now, Merging
